compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala

```diff
@@ -46,14 +63,141 @@ class ExtractSemanticDB extends Phase:
   // Check not needed since it does not transform trees
   override def isCheckable: Boolean = false
 
-  override def run(using Context): Unit =
-    val unit = ctx.compilationUnit
-    val extractor = Extractor()
-    extractor.extract(unit.tpdTree)
-    ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
+  override def runOn(units: List[CompilationUnit])(using ctx: Context): List[CompilationUnit] = {
+    val sourceRoot = ctx.settings.sourceroot.value
+    val appendDiagnostics = phaseMode == ExtractSemanticDB.PhaseMode.AppendDiagnostics
+    if (appendDiagnostics)
+      val warnings = ctx.reporter.allWarnings
+        .groupBy(w => w.pos.source)
+        .map((k, ws) => (k, ws.map(_.toSemanticDiagnostic)))
+        .toMap
+      units.map { unit =>
+        val unitCtx = ctx.fresh.setCompilationUnit(unit).withRootImports
+        val outputDir =
+          ExtractSemanticDB.semanticdbPath(
+            unit.source,
+            ExtractSemanticDB.outputDirectory(using unitCtx),
+            sourceRoot
+          )
+        val source = unit.source
+        (outputDir, source)
+      }.asJava.parallelStream().forEach { case (out, source) =>
+        warnings.get(source).foreach { ws =>
+          ExtractSemanticDB.appendDiagnostics(source, ws, out)
+        }
+      }
+    else
+      units.foreach { unit =>
+        val outputDir =
+          ExtractSemanticDB.semanticdbPath(
+            unit.source,
+            ExtractSemanticDB.outputDirectory(using ctx.fresh.setCompilationUnit(unit).withRootImports),
+            sourceRoot
+          )
+        val extractor = ExtractSemanticDB.Extractor()
+        extractor.extract(unit.tpdTree)
```

Comment:
not quite sure but there could be a slight difference here given that the extractor isn't running with `unitCtx` like before